src/kestrel/codegen/relations.py

```diff
@@ -230,52 +230,122 @@ def _generate_paramstix_comparison_expressions(
 def fine_grained_relational_process_filtering(
     local_var, prefetch_entity_table, store, config
 ):
+    # Two-step search for matched processes
+    # 1. pivot process search (find pivot process in fil_processes against ref_processes)
+    # 2. precise process search (find process in fil_processes against pivot_processes)
+
+    # two situations worth mentioning:
+    # - in Linux, a new process will be forked, then exec to change name. In this case,
+    #   we need to search for pivot_rows to identify process with even name changed,
+    #   then get all records of both process before name change and after name change.
+    # - ppid is useful to identify a process with pid, however, in one situation, the
+    #   ppid data is not available in the first phase of FIND (creating of local_var
+    #   using deref in firepit)---FIND parent process of current process. This is because
+    #   most datasource does not store *parent parent process pid* for deref to get ppid
+    #   of the parent. In this case, we need to search for pivot_rows to infer the ppid.
 
     _logger.debug(
         f"start fine-grained relational process filtering for prefetched table: {prefetch_entity_table}"
     )
 
-    query_ref = Query(
-        [
-            Table(local_var.entity_table),
-            Join("__contains", "id", "=", "target_ref"),
-            Join("observed-data", "source_ref", "=", "id"),
-            Projection(["pid", "name", "first_observed", "last_observed"]),
-        ]
+    ref_processes = _query_process_with_time_and_ppid(store, local_var.entity_table)
+
+    fil_processes = _query_process_with_time_and_ppid(store, prefetch_entity_table)
```

Comment:
What are `ref_processes` and `fil_processes`?

Reply:
the logic is a little complex, comments added.